### PR TITLE
ISPN-6977 ISPN-6983 ISPN-6984 Random test failures

### PR DESCRIPTION
**Pull request by danberindei**
https://issues.jboss.org/browse/ISPN-6977
https://issues.jboss.org/browse/ISPN-6983
https://issues.jboss.org/browse/ISPN-6984
https://issues.jboss.org/browse/ISPN-5424

ISPN-6977 NumOwnersNodeCrashInSequenceTest random failures
ISPN-6983 ThreeNodesSplitAndMergeTest random failures
ISPN-6984 ConcurrentJoinTest random failures
ISPN-5424 Replace SemaphoreCompletionService with LimitedExecutor

### CODE REVIEW

File: core/src/main/java/org/infinispan/executors/LimitedExecutor.java [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+/**
+ * Executes tasks in the given executor, but never has more than {@code maxConcurrentTasks} tasks running at the same
+ * time.
+ *
+ * The number of concurrent tasks can be tweaked at any time with {@link #addPermit()} and {@link #removePermit()}.
+ *
+ * @author Dan Berindei
+ * @since 9.0
+ */
+public class LimitedExecutor implements Executor {
```
Comment (slaskawi) [inline core/src/main/java/org/infinispan/executors/LimitedExecutor.java]: How about using CompletionService with ThreadPoolExecutor? Because if I understand the idea of this class correctly, we want to execute X tasks concurrently and store results in a list/queue?

File: core/src/test/java/org/infinispan/executors/LimitedExecutorTest.java [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+      LimitedExecutor limitedExecutor = new LimitedExecutor(NAME, executor, 1);
+
+      CompletableFuture<Void> blockingFuture = new CompletableFuture<>();
+      CompletableFuture<String> cf1 = new CompletableFuture<>();
+      limitedExecutor.executeAsync(() -> blockingFuture.thenRunAsync(() -> cf1.complete("bla")));
+
+      CompletableFuture<String> cf2 = new CompletableFuture<>();
+      limitedExecutor.execute(() -> cf2.complete("bla"));
+
+      Thread.sleep(100);
```
Comment (slaskawi) [inline core/src/test/java/org/infinispan/executors/LimitedExecutorTest.java]: Can we get rid of those sleeps somehow?

File: core/src/main/java/org/infinispan/executors/LimitedExecutor.java (hunk 1 above)
Comment (danberindei) [inline core/src/main/java/org/infinispan/executors/LimitedExecutor.java]: No, `SemaphoreCompletionService` stores the results in a queue, but `LimitedExecutor` doesn't keep track of results.

File: core/src/test/java/org/infinispan/executors/LimitedExecutorTest.java (hunk 2 above)
Comment (danberindei) [inline core/src/test/java/org/infinispan/executors/LimitedExecutorTest.java]: I can change them to `sleep(10)`, but there's no deterministic way to check that `cf2` does not complete before `blockingFuture`.

Comment (rvansa) [inline core/src/test/java/org/infinispan/executors/LimitedExecutorTest.java]: @danberindei How about incrementing shared AtomicInteger and checking that the first lambda gets 1 and second gets 2?

File: core/src/main/java/org/infinispan/executors/LimitedExecutor.java (hunk 1 above)
Comment (slaskawi) [inline core/src/main/java/org/infinispan/executors/LimitedExecutor.java]: So why can't we use a pure Executor than (if we don't about the results)?

Comment (danberindei) [inline core/src/main/java/org/infinispan/executors/LimitedExecutor.java]: Because I still want to limit the number of tasks that run concurrently. A `LimitedExecutor(1)` allows me to submit 2 tasks and runs only 1 at a time, but without creating a new thread for each `LimitedExecutor` instance. `executeAsync` is also something you don't get in a regular `Executor` :)

Comment (slaskawi) [inline core/src/main/java/org/infinispan/executors/LimitedExecutor.java]: Ok understood. Thanks for clarification.

**Comment by slaskawi** [comment]
LGTM, @pruivo, @rvansa ?

File: core/src/main/java/org/infinispan/executors/LimitedExecutor.java [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+
+import org.infinispan.util.logging.Log;
+import org.infinispan.util.logging.LogFactory;
+import org.jboss.logging.NDC;
+
+/**
+ * Executes tasks in the given executor, but never has more than {@code maxConcurrentTasks} tasks running at the same
+ * time.
+ *
+ * The number of concurrent tasks can be tweaked at any time with {@link #addPermit()} and {@link #removePermit()}.
```
Comment (rvansa) [inline core/src/main/java/org/infinispan/executors/LimitedExecutor.java]: `addPermit` and `removePermit` are private - how can you tweak it?

File: core/src/main/java/org/infinispan/statetransfer/InboundTransferTask.java [hunk 4]
```diff
@@ -48,27 +50,9 @@
...
-    */
-   private volatile boolean isCompletedSuccessfully = false;
-
-   /**
     * This latch is counted down when all segments are completely received or in case of task cancellation.
     */
-   private final CountDownLatch completionLatch = new CountDownLatch(1);
-
-   private final StateConsumerImpl stateConsumer;
+   private final CompletableFuture<Void> completionLatch = new CompletableFuture<>();
```
Comment (rvansa) [inline core/src/main/java/org/infinispan/statetransfer/InboundTransferTask.java]: update field name

File: core/src/main/java/org/infinispan/statetransfer/InboundTransferTask.java [hunk 5]
```diff
@@ -34,12 +34,14 @@
    private static final boolean trace = log.isTraceEnabled();
 
    /**
-    * All access to fields {@code segments} and {@code finishedSegments} must bead done while synchronizing on {@code segments}.
+    * All access to fields {@code segments} and {@code finishedSegments} must bead done while synchronizing on {@code
```
Comment (rvansa) [inline core/src/main/java/org/infinispan/statetransfer/InboundTransferTask.java]: Next time please do all the reformatting in separate commit, for easier review.

**Comment by rvansa** [comment]
Generally OK, though the formatting changes are annoying. Does checkstyle enforce them?

**Comment by slaskawi** [comment]
@danberindei Once the questions are answered and @rvansa is happy, I'll integrate it.

File: core/src/test/java/org/infinispan/executors/LimitedExecutorTest.java [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+      CompletableFuture<String> cf1 = new CompletableFuture<>();
+      limitedExecutor.execute(() -> {
+         awaitLatch(latch);
+         cf1.complete("bla");
+      });
+
+      CompletableFuture<String> cf2 = new CompletableFuture<>();
+      limitedExecutor.execute(() -> cf2.complete("bla"));
+
+      Thread.sleep(10);
```
Comment (slaskawi) [inline core/src/test/java/org/infinispan/executors/LimitedExecutorTest.java]: Can we use `Thread#yield` here? Just to let JVM continue running other thread?

File: core/src/main/java/org/infinispan/statetransfer/StateConsumerImpl.java [hunk 7]
```diff
@@ -944,50 +951,44 @@ private void removeStaleData(final Set<Integer> removedSegments) throws Interrup
...
          try {
-            InvalidateCommand invalidateCmd = commandsFactory.buildInvalidateCommand(EnumUtil.bitSetOf(CACHE_MODE_LOCAL, SKIP_LOCKING), keysToRemove.toArray());
+            InvalidateCommand invalidateCmd = commandsFactory
+                  .buildInvalidateCommand(EnumUtil.bitSetOf(CACHE_MODE_LOCAL, SKIP_LOCKING), keysToRemove.toArray());
             InvocationContext ctx = icf.createNonTxInvocationContext();
             ctx.setLockOwner(invalidateCmd.getKeyLockOwner());
             interceptorChain.invoke(ctx, invalidateCmd);
 
-            if (trace) log.tracef("Removed %d keys, data container now has %d keys", keysToRemove.size(), dataContainer.size());
+            if (trace)
```
Comment (slaskawi) [inline core/src/main/java/org/infinispan/statetransfer/StateConsumerImpl.java]: Some time ago I complained about those constructs (`if(trace) log.tracef(...)`). What if someone changes the logging level in the runtime (via JMX for example)? This seems to be a normal thing to do...

I know it's a common practice to use this `if(trace)` things... but on the other hand I believe it is wrong...

Comment (rvansa) [inline core/src/main/java/org/infinispan/statetransfer/StateConsumerImpl.java]: I think that performance tradeoff is more desirable than ability to change runtime level to TRACE. If you could demonstrate that this does not have any perf effect on modern JVMs by a stress test, the practice could be reconsidered.

Comment (slaskawi) [inline core/src/main/java/org/infinispan/statetransfer/StateConsumerImpl.java]: Ok, sounds like a plan :)

File: core/src/test/java/org/infinispan/executors/LimitedExecutorTest.java (hunk 6 above)
Comment (danberindei) [inline core/src/test/java/org/infinispan/executors/LimitedExecutorTest.java]: We could, but then the test may not catch all situations where `LimitedExecutor` runs the task too soon. So I'd rather keep the `sleep(10)`.

File: core/src/main/java/org/infinispan/statetransfer/StateConsumerImpl.java (hunk 7 above)
Comment (danberindei) [inline core/src/main/java/org/infinispan/statetransfer/StateConsumerImpl.java]: @slaskawi You can try, but I doubt it -- the difference is always going to be visible, especially for fast operations like replicated mode reads.

**Comment by slaskawi** [comment]
Looking

**Comment by slaskawi** [comment]
The PR & master builds have lots of failures... Let's give this one a go...
